grass.script: Split the command string using shell-like syntax on the win32 platform same as on POSIX-compliant platforms

tmszi opened the pull request:
Closes #1063.

**To Reproduce**
Steps to reproduce the behavior:

1. Launch GRASS GIS
2. Launch wx0 monitor `d.mon start=wx0`
3. Try display some vector map with SQL query (where parameter), don't forget escaping MS Windows Command Propmt special characters `< > ( ) & | , ; "` with `^` carret symbol please, e.g. `d.vect map=census where="cat ^> 10 AND cat ^< 20"`


**Additional context**

More info is in [commit message](https://github.com/OSGeo/grass/commit/573cd906e0162c63d62a6dc218bfa062023b5d57).

hellik commented:
issue confirmed

hellik commented:
>don't forget escaping MS Windows Command Propmt >special characters < > ( ) & | , ; " with ^ carret symbol >please, e.g. d.vect map=census where="cat ^> 10 AND >cat ^< 20"

But why is

**cat > 10** 

working without escaping?

tmszi commented:
> > don't forget escaping MS Windows Command Propmt >special characters < > ( ) & | , ; " with ^ carret symbol >please, e.g. d.vect map=census where="cat ^> 10 AND >cat ^< 20"
> 
> But why is
> 
> **cat > 10**
> 
> working without escaping?

Yes you are right, because `where=cat > 10` is in Python splitted as `where=cat` (and  `> 10` is stripped) , and therefore d.vect display **all categories** (not `cat > 10`). You can verify this on the wx0 monitor, but better example is `where=cat < 10`.

wenzeslaus reviewed (commented):
Simpler code, aligned behavior across platforms (GUI Console is not system terminal), so I general yes.

I would appreciate more explanatory description (that is the final commit message and possibly also the PR description). Although I think I got the idea, I don't really get from the description what are the desired, current, and failing cases.

tmszi commented:
Rebase https://github.com/OSGeo/grass/commit/0772c39616d2cea0f926d6532a596b605e3dddc.

This is a significant fix and should be merged.

wenzeslaus reviewed (commented):
I'm for merging this, but I'm still not quite able to put all the pieces together.

When it happens and how the carret fits into this? The issue happens when using d.mon in terminal. What happens with d.vect in Console in GUI? Where is this function used and what strings it processes?

Is the function documentation still valid? The PR title would suggest it is not. The implementation has slightly different behavior for Windows but without explanation.

A test would at least clarify the intended behavior which would help clarifying the intention further.

I'm sorry for all the questions, but I guess insufficient documentation is what lead to the bad usage of this function in the first place.

wenzeslaus commented:
A test for this could go into [python/grass/script/tests](https://github.com/OSGeo/grass/tree/main/python/grass/script/tests) and it is a great candidate for using pytest.

tmszi commented:
> When it happens and how the carret fits into this? The issue happens when using d.mon in terminal. What happens with d.vect in Console in GUI? Where is this function used and what strings it processes?

First read issue #1063 carefully please.

**MS Windows platform Command Prompt CMD**:

Display some vector map e.g. census and use `where` parameter:

A. Default behavior:

I did not find anywhere in the `d.vect` documentation manual that it is **necessary to escape special characters** < > ( ) & | , ; " **with the `^`** carret character when command is **runned from CMD only**. That's why I update doc of `d.vect` module.

Without carret symbol (d.vect module fail with error message):

```
# Launch display
d.mon start=wx0

# Display some vector map e.g. census and use where parameter
C:\Users\User>d.vect census where="cat < 10"
The system cannot find the path specified.
```

With carret char (d.vect module fail again with error message):

```
C:\Users\User>d.vect census where="cat ^< 10"
WARNING: Illegal filename <10">. Character <"> not allowed.
ERROR: Vector map <10"> not found

C:\Users\User>Syntax error: No closing quotationTraceback (most recent call last):
  File "C:\Program Files\GRASS GIS 8.4\Python39\lib\site-packages\wx\core.py", line 2346, in Notify
    self.notify()
  File "C:\Program Files\GRASS GIS 8.4\gui\wxpython\mapdisp\main.py", line 628, in watcher
    self.mapDisplay.GetMap().GetLayersFromCmdFile()
  File "C:\Program Files\GRASS GIS 8.4\gui\wxpython\mapdisp\main.py", line 191, in GetLayersFromCmdFile
    ltype = utils.command2ltype[cmd[0]]
IndexError: list index out of range
```

Internaly is commandline parsed in Python with following function (under OS MS Windows is used parameter `posix=False`, which is the source of  incorrect parsing of the `where` parameter of d.vect module):

https://github.com/OSGeo/grass/blob/5deae3aaa1c538f0b0e18eb2b2ce39169e33fbd9/python/grass/script/utils.py#L308-L310

parsed result list:

```python
['d.vect', 'map="census"', 'layer="1"', 'display="shape"', 'type="point,line,area,face"', 'where="cat', '<', '10"', 'color="0:29:57"', 'fill_color="0:103:204"', 'width=0', 'width_scale=1', 'icon="basic/x"', 'size=5', 'icon_area="legend/area"', 'icon_line="legend/line"', 'label_layer="1"', 'label_color="red"', 'label_bgcolor="none"', 'label_bcolor="none"', 'label_size=8', 'xref="left"', 'yref="center"']
```

Look at how is parsed `where` param: ```'where="cat', '<', '10"'``` which is **incorrect**.

If function `return shlex.split(s, posix=(sys.platform != "win32")) ` param `posix` argument is replaced with `True` on OS MS Windows `return shlex.split(s.replace("\\", r"\\")) ` parsed commandline `d.vect`  module `where` param argument is parsed correctly as  `'where=cat < 10'`

Expected behavior (solved with this PR):

```
C:\Users\User>d.vect census where="cat ^< 10"
d.vect complete.
```
 
**MS Windows platform wxGUI** console is not affected with changes in this PR and reported issue.

Is such an explanation sufficient to understand the problem and solve it?

> Is the function documentation still valid? The PR title would suggest it is not. The implementation has slightly different behavior for Windows but without explanation.

Yes I agree func doc is not valid and explanation on MS Windows platform is needed too. 

> A test would at least clarify the intended behavior which would help clarifying the intention further.

if possible, I will add a test.

wenzeslaus commented:
Thank you so much. With your new description, it is clear to me now. I see that the caret symbol is really a side issue here as the issue is always passing display commands as posix somewhere in d.mon the rendering machinery while expecting them as platform specific. In this context, I'm now curious why the backslashes need to be doubled on Windows but don't on Linux.  You comment also clarifies why the split is not relevant to v.extract.

I agree with the addition of discussion of the caret symbol into the documentation of d.vect. It applies to v.extract as well as anything else, though, because it is a CMD thing, right? But that's an issue for later.

tmszi commented:
> I'm now curious why the backslashes need to be doubled on Windows but don't on Linux. 

During my current testing, on the MS Windows 11, I found that this is not necessary, but  I don't know reason why I added it (PR is from 2021). Originally I tested PR on MS Windows 10.

> You comment also clarifies why the split is not relevant to v.extract

v.extract module doesn't use Python `shlex.split()` func.

petrasovaa commented on python/grass/script/utils.py:
```diff
@@ -306,8 +305,27 @@ def get_num_suffix(number, max_number):
 def split(s):
-    """!Platform specific shlex.split"""
-    return shlex.split(s, posix=(sys.platform != "win32"))
+    """Same shlex.split() func on all OS platforms
+
+    We don't use parameter posix=True on the OS MS Windows due incorrectly
+    splliting commandlines parameters.
+
+    e.g. d.vect where="cat < 10"
+
+    Is splitted incorectly as follow:
+
```
```suggestion
    """Same shlex.split() func on all OS platforms

    We don't use parameter posix=True on the OS MS Windows due to incorrectly
    splitting command line parameters:

    e.g. d.vect where="cat < 10"

    is split incorrectly as follows:

```

petrasovaa commented:
Just to give an overview where this is used:
```
./scripts/d.redraw/d.redraw.py:27:from grass.script.utils import split
```
```
./gui/wxpython/mapdisp/main.py:149:                    if "=" in utils.split(dWhatCmd)[1]:
./gui/wxpython/mapdisp/main.py:150:                        maps = utils.split(dWhatCmd)[1].split("=")[1].split(",")
./gui/wxpython/mapdisp/main.py:152:                        maps = utils.split(dWhatCmd)[1].split(",")
./gui/wxpython/mapdisp/main.py:184:                cmd = utils.split(line.strip())
./gui/wxpython/core/gconsole.py:473:        :param command: command given as a list (produced e.g. by utils.split())
./gui/wxpython/core/gconsole.py:475:        e.g. by utils.split()) and key 'cmdString' with original cmd
./gui/wxpython/gmodeler/dialogs.py:292:            cmd = utils.split(str(line))
./gui/wxpython/gui_core/prompt.py:133:            cmd = utils.split(str(cmdString))
./gui/wxpython/gui_core/prompt.py:135:            cmd = utils.split(EncodeString((cmdString)))
./gui/wxpython/gui_core/prompt.py:393:            splitted = utils.split(str(entry))
./gui/wxpython/gui_core/forms.py:3243:        cmd = utils.split(sys.argv[1])
./gui/wxpython/gui_core/menu.py:106:                cmd = utils.split(str(command))
./gui/wxpython/gui_core/menu.py:108:                cmd = utils.split(EncodeString((command)))
```

landam commented:
I am able to reproduce this issue on Windows in wxGUI Python tab:

```
from grass.script.utils import split
split('d.vect census where="cat < 10"')
['d.vect', 'census', 'where="cat', '<', '10"']
import shlex
shlex.split('d.vect census where="cat < 10"')
['d.vect', 'census', 'where=cat < 10']
```
